File: CSES Problem Set/Advanced Techniques/Meet in the Middle.py
N, x = map(int, input().split())
t = list(map(int, input().split()))
t.sort()

if N == 1:
    print(1 if x == t[0] else 0)
else:
    freq = {}
    for i in range(1 << (N // 2 - 1)):
        sum_val = 0
        for j in range(N // 2 - 1):
            if i & (1 << j):
                sum_val += t[j]
        freq[sum_val] = freq.get(sum_val, 0) + 1

    cnt = 0
    for i in range(1 << ((N + 1) // 2 + 1)):
        sum_val = 0
        for j in range((N + 1) // 2 + 1):
            if i & (1 << j):
                sum_val += t[N // 2 - 1 + j]
        cnt += freq.get(x - sum_val, 0)

    print(cnt)

File: CSES Problem Set/Advanced Techniques/Meet in the Middle.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
const int maxN = 40;

int N, t[maxN];
ll x, sum, cnt;
unordered_map<ll, int> freq;

int main()
{
    cin >> N >> x;
    for (int i = 0; i < N; i++)
        cin >> t[i];

    sort(t, t + N);

    if (N == 1)
    {
        cout << ((x == t[0]) ? 1 : 0) << endl;
        return 0;
    }

    freq.reserve(1 << (N / 2 - 1));

    for (int i = 0; i < (1 << (N / 2 - 1)); i++)
    {
        sum = 0;
        for (int j = 0; j < N / 2 - 1; j++)
            if (i & (1 << j))
                sum += t[j];
        freq[sum]++;
    }

    for (int i = 0; i < (1 << ((N + 1) / 2 + 1)); i++)
    {
        sum = 0;
        for (int j = 0; j < (N + 1) / 2 + 1; j++)
            if (i & (1 << j))
                sum += t[N / 2 - 1 + j];
        if (freq.find(x - sum) != freq.end())
            cnt += freq[x - sum];
    }

    cout << cnt << endl;
}